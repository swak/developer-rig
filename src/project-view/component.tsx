import * as React from 'react';
import './component.sass';
import { RigProject } from '../core/models/rig';
import { fetchHostingStatus, startBackend, hostFrontend, startFrontend, stopHosting, StopOptions } from '../util/api';
import { fetchUserExtensionManifest } from '../util/extension';

export interface ProjectViewProps {
  rigProject: RigProject,
  userId: string;
  onChange: (rigProject: RigProject) => void,
}

enum HostingResult {
  None = '',
  NotRunning = 'not running',
  Started = 'started',
  Running = 'running',
  Exited = 'exited',
}

interface State {
  backendResult: string;
  frontendResult: string;
}

export class ProjectView extends React.Component<ProjectViewProps, State>{
  public state: State = {
    backendResult: HostingResult.None,
    frontendResult: HostingResult.None,
  };

  constructor(props: ProjectViewProps) {
    super(props);
    fetchHostingStatus().then((status) => {
      const { rigProject } = props;
      this.setState({
        backendResult: getStatus(rigProject.backendCommand, status.isBackendRunning),
        frontendResult: getStatus(rigProject.frontendFolderName || rigProject.frontendCommand, status.isFrontendRunning),
      });

      function getStatus(value: string, isRunning: boolean): HostingResult {
        return value ? isRunning ? HostingResult.Running : HostingResult.NotRunning : HostingResult.None;
      }
    });
  }

  public onChange = (input: React.FormEvent<HTMLInputElement>) => {
    const { name, value } = input.currentTarget;
    if (name === 'name') {
      if (this.props.rigProject.isLocal) {
        const manifest = Object.assign({}, this.props.rigProject.manifest, { name: value });
        this.props.onChange({ manifest } as any as RigProject);
      }
    } else {
      this.props.onChange({ [name]: value } as any as RigProject);
    }
  }

  private toggleBackend = async () => {
    const rigProject = this.props.rigProject;
    if (rigProject.backendCommand) {
      try {
        this.setState({ backendResult: HostingResult.None });
        if (this.getIsRunning(this.state.backendResult)) {
          const { backendResult } = await stopHosting(StopOptions.Backend);
          this.setState({ backendResult });
        } else {
          await startBackend(rigProject.backendCommand, rigProject.projectFolderPath);
          this.setState({ backendResult: HostingResult.Running });
        }
      } catch (ex) {
        this.setState({ backendResult: ex.message });
      }
    }
  }

  private toggleFrontend = async () => {
    const rigProject = this.props.rigProject;
    if (rigProject.frontendFolderName || rigProject.frontendCommand) {
      try {
        const frontendState = this.state.frontendResult;
        this.setState({ frontendResult: HostingResult.None });
        if (this.getIsRunning(frontendState)) {
          const { frontendResult } = await stopHosting(StopOptions.Frontend);
          this.setState({ frontendResult });
        } else if (rigProject.frontendCommand) {
          await startFrontend(rigProject.frontendFolderName, rigProject.frontendCommand, rigProject.projectFolderPath);
          this.setState({ frontendResult: HostingResult.Started });
        } else {
          let frontendPort: number;
          ['panel', 'component', 'videoOverlay', 'mobile', 'config', 'liveConfig'].some((name) => {
            const view = (rigProject.manifest.views as any)[name];
            if (view && view.viewerUrl) {
              const url = new URL(view.viewerUrl);
              frontendPort = parseInt(url.port, 10) || (url.protocol === 'http:' ? 80 : 443);
              return true;
            }
            return false;
          });
          if (!frontendPort) {
            throw new Error('Cannot determine front-end port from extension');
          }
          await hostFrontend(rigProject.frontendFolderName, rigProject.isLocal, frontendPort, rigProject.projectFolderPath);
          this.setState({ frontendResult: HostingResult.Started });
        }
      } catch (ex) {
        this.setState({ frontendResult: ex.message });
      }
    }
  }

  private getExtensionViews(rigProject: RigProject): string {
    if (rigProject.manifest.views) {
      const extensionViewTypes = ['panel', 'component', 'videoOverlay', 'mobile'];
      return ['Panel', 'Component', 'Video Overlay', 'Mobile'].filter((_, index) => {
        return Object.getOwnPropertyDescriptor(rigProject.manifest.views, extensionViewTypes[index]);
      }).join(', ');
    }
    return '';
  }

  private viewDocumentation() {
    window.open('https://dev.twitch.tv/docs/extensions/', 'developer-rig-help');
  }

  private viewTutorial() {
    window.open('https://www.twitch.tv/videos/239080621', 'developer-rig-help');
  }

  private getIsRunning(result: string) {
    return result === HostingResult.Started || result === HostingResult.Running;
  }

  private refreshManifest = async () => {
    const { secret, manifest: { id: clientId, version } } = this.props.rigProject;
    const manifest = await fetchUserExtensionManifest(false, this.props.userId, secret, clientId, version);
    this.props.onChange({ manifest } as any as RigProject);
  }

  public render() {
    const rigProject = this.props.rigProject;
    const frontendHostText = this.getIsRunning(this.state.frontendResult) ? 'Stop Hosting' : 'Host with Rig';
    const frontendCommandText = this.getIsRunning(this.state.frontendResult) ? 'Deactivate' : 'Activate';
    const backendText = this.getIsRunning(this.state.backendResult) ? 'Deactivate' : 'Activate';
    return (
      <div className="project-view">
        <div className="project-view__section project-view__section--left">
          <label className="project-view-property">
            <div className="project-view-property__name">Project Name</div>
            <input className="project-view-property__input project-view-property__input--half" type="text" name="name" value={rigProject.manifest.name} onChange={this.onChange} />
          </label>
          <label className="project-view-property" title="This is the path to your front-end files relative to the Project Folder.  If there is no Project Folder, ensure this path is absolute.">
            <div className="project-view-property__name">Front-end Files Location</div>
            <input className="project-view-property__input" type="text" name="frontendFolderName" value={rigProject.frontendFolderName} onChange={this.onChange} />
            {!rigProject.frontendCommand && (
              <>
                <button className="project-view__button" title="" onClick={this.toggleFrontend}>{frontendHostText}</button>
                <div title="This is the result of the front-end hosting command." className="project-view-property__result">{this.state.frontendResult}</div>
              </>
            )}
          </label>
          <label className="project-view-property" title="This is the command used to host your front-end files.  If you leave this blank, you may use the Developer Rig to host your front-end files.">
            <div className="project-view-property__name">Front-end Host Command</div>
            <input className="project-view-property__input" type="text" name="frontendCommand" value={rigProject.frontendCommand} onChange={this.onChange} placeholder="Leave blank for Developer Rig hosting" />
            {rigProject.frontendCommand && (
              <>
                <button className="project-view__button" title="" onClick={this.toggleFrontend}>{frontendCommandText}</button>
                <div title="This is the result of the front-end activation command." className="project-view-property__result">{this.state.frontendResult}</div>
              </>
            )}
          </label>
          <label className="project-view-property" title="This is the command used to run your back-end.  If there is a Project Folder, this command is run with that folder as its current directory.">
            <div className="project-view-property__name">Back-end Run Command</div>
            <input className="project-view-property__input" type="text" name="backendCommand" value={rigProject.backendCommand} onChange={this.onChange} />
            <button className="project-view__button" title="" onClick={this.toggleBackend}>{backendText}</button>
            <div title="This is the result of the back-end activation command." className="project-view-property__result">{this.state.backendResult}</div>
          </label>
          <label className="project-view-property">
            <div className="project-view-property__name">Project Folder</div>
            <input className="project-view-property__input" type="text" name="projectFolderPath" value={rigProject.projectFolderPath} onChange={this.onChange} />
          </label>
          <label className="project-view-property">
            <div className="project-view-property__name">Extension Types</div>
            <div className="project-view-property__value">{this.getExtensionViews(rigProject)}</div>
          </label>
          {!rigProject.isLocal && (
            <>
              <label className="project-view-property">
                <button className="project-view__button project-view__button--first" onClick={this.refreshManifest}>Refresh Manifest</button>
              </label>
              <label className="project-view-property">
                Make sure your Asset Paths are correct. Your "Testing Base URI" must
                be https://localhost.rig.twitch.tv:8080/ to use the Developer Rig.
                Also, ensure the individual Extension Type paths are correct.
                Click <a href={`https://dev.twitch.tv/projects/${rigProject.manifest.id}/files`} target="dev-site">here</a> then
                click "Manage" for version {rigProject.manifest.version} then
                click "Asset Hosting" to check and adjust if needed.
              </label>
            </>
          )}
        </div>
        <div className="project-view__vertical-bar" />
        <div className="project-view__section project-view__section--right">
          <div className="project-view__title">How to Run an Extension in the Rig</div>
          <ol>
            <li className="project-view__item">
              <div className="project-view__item-text project-view__item-text--title">Host your front-end files.</div>
              <div className="project-view__item-text">You can host your front-end files with the Rig by entering the path to
                your HTML files in the "Front-end Files Location" text box and clicking the "Host with Rig" button.  If your
                front-end requires a custom command, such as for a React application, enter its activation command in the
                "Front-end Host Command" text box and click the "Activate" button.</div>
            </li>
            <li className="project-view__item">
              <div className="project-view__item-text project-view__item-text--title">If your extension has a back-end, run it locally.</div>
              <div className="project-view__item-text">You can run your back-end service from the Rig by entering the command
                to activate it in the "Back-end Run Command" text box and clicking the "Activate" button.</div>
            </li>
            <li className="project-view__item">
              <div className="project-view__item-text project-view__item-text--title">Go to Extension Views and add the
                Extension Types that match your extension.</div>
            </li>
          </ol>
          <button className="project-view__button project-view__button--first" onClick={this.viewTutorial}>View Tutorial</button>
          <button className="project-view__button" onClick={this.viewDocumentation}>Go to Documentation</button>
        </div>
      </div>
    );
  }
}
