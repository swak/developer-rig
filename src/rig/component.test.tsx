import { setupShallowTest } from '../tests/enzyme-util/shallow';
import { createViewsForTest, createExtensionManifestForTest } from '../tests/constants/extension';
import { mockFetchForUserInfo } from '../tests/mocks';
import { NavItem } from '../constants/nav-items';
import { RigComponent } from './component';
import { ExtensionAnchors } from '../constants/extension-types';
import { ViewerTypes } from '../constants/viewer-types';
import { RigExtensionView, RigProject } from '../core/models/rig';
import { ExtensionViewDialogState } from '../extension-view-dialog';
import { ExtensionAnchor, ExtensionViewType } from '../constants/extension-coordinator';

let globalAny = global as any;

localStorage.setItem('projects', '[{},{}]');

const setupShallow = setupShallowTest(RigComponent, () => ({
  session: { displayName: 'test', login: 'test', id: 'test', profileImageUrl: 'test.png', authToken: 'test' },
  saveManifest: jest.fn(),
  userLogin: jest.fn()
}));

describe('<RigComponent />', () => {
  function setUpProjectForTest(type: ExtensionAnchor) {
    const extensionViews = createViewsForTest(1, ExtensionAnchors[type], ViewerTypes.LoggedOut);
    localStorage.setItem('projects', JSON.stringify([{ extensionViews }]));
  }

  it('renders correctly', () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders extension view correctly', () => {
    setUpProjectForTest(ExtensionAnchor.Panel);
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
    expect((wrapper.find('ExtensionViewContainer') as any).props().extensionViews).toHaveLength(1);
  });

  it('gets extension views from local storage correctly', () => {
    setUpProjectForTest(ExtensionAnchor.Panel);
    const testViews = createViewsForTest(1, ExtensionAnchors[ExtensionAnchor.Panel], ViewerTypes.LoggedOut);
    const { wrapper } = setupShallow();
    const instance = wrapper.instance() as RigComponent;
    expect(instance.state.currentProject.extensionViews).toEqual(testViews);
  });

  it('deletes extension view correctly', () => {
    setUpProjectForTest(ExtensionAnchor.Panel);
    const { wrapper } = setupShallow();
    const instance = wrapper.instance() as RigComponent;
    expect((wrapper.find('ExtensionViewContainer') as any).props().extensionViews).toHaveLength(1);

    instance.deleteExtensionView('1');
    wrapper.update();
    expect((wrapper.find('ExtensionViewContainer') as any).props().extensionViews).toHaveLength(0);
  });

  it('toggles state when edit dialog is opened/closed', () => {
    setUpProjectForTest(ExtensionAnchor.Component);
    const { wrapper } = setupShallow();
    const instance = wrapper.instance() as RigComponent;

    instance.openEditViewHandler('1');
    expect(instance.state.showingEditView).toBe(true);
    expect(instance.state.idToEdit).toBe('1');

    instance.closeEditViewHandler();
    expect(instance.state.showingEditView).toBe(false);
    expect(instance.state.idToEdit).toBe('0');
  });

  it('edit changes the view and sets them correctly', () => {
    setUpProjectForTest(ExtensionAnchor.Component);
    const { wrapper } = setupShallow();
    const instance = wrapper.instance() as RigComponent;

    instance.openEditViewHandler('1');
    expect(instance.state.showingEditView).toBe(true);
    expect(instance.state.idToEdit).toBe('1');

    instance.editViewHandler({ x: 25, y: 25 });

    const views = instance.state.currentProject.extensionViews;
    const editedView = views.filter((element: RigExtensionView) => element.id === '1');
    expect(editedView[0].x).toEqual(25);
    expect(editedView[0].y).toEqual(25);
    expect(instance.state.showingEditView).toBe(false);
    expect(instance.state.idToEdit).toBe('0');
  });

  it('correctly toggles state when create extension view is opened/closed', () => {
    setUpProjectForTest(ExtensionAnchor.Panel);
    const { wrapper } = setupShallow();
    const instance = wrapper.instance() as RigComponent;
    instance.state.currentProject = { manifest: createExtensionManifestForTest() } as RigProject;

    instance.openExtensionViewHandler();
    expect(instance.state.showingExtensionsView).toBe(true);

    instance.closeExtensionViewDialog();
    expect(instance.state.showingExtensionsView).toBe(false);
  });

  it('correctly sets state when viewHandler is invoked', () => {
    setUpProjectForTest(ExtensionAnchor.Panel);
    const { wrapper } = setupShallow();
    wrapper.setState({ currentProject: { manifest: createExtensionManifestForTest() } as RigProject });

    const instance = wrapper.instance() as RigComponent;
    instance.viewerHandler(NavItem.ExtensionViews);
    expect(instance.state.selectedView).toBe(NavItem.ExtensionViews);
  });

  it('gets the correct views when getExtensionViews invoked', () => {
    const testViews = createViewsForTest(1, ExtensionAnchors[ExtensionAnchor.Panel], ViewerTypes.LoggedOut);
    setUpProjectForTest(ExtensionAnchor.Panel);
    const { wrapper } = setupShallow();
    const instance = wrapper.instance() as RigComponent;

    expect(instance.state.currentProject.extensionViews).toEqual(testViews);
  });

  it('returns correct data for mobile ', () => {
    const { wrapper } = setupShallow();
    const testDialogState = {
      width: 0,
      height: 0,
      frameSize: 'iPhone X (375x822)',
      extensionViewType: ExtensionViewType.Mobile
    } as ExtensionViewDialogState;
    const expectedMobileFrameSize = {
      width: 375,
      height: 822,
    };
    const instance = wrapper.instance() as RigComponent;

    let frameSize = instance.getFrameSizeFromDialog(testDialogState);
    expect(frameSize).toEqual(expectedMobileFrameSize);
  });

  it('returns correct data for other types', () => {
    const { wrapper } = setupShallow();
    const overlayTestDialogState = {
      width: 0,
      height: 0,
      frameSize: '640x480',
      extensionViewType: ExtensionViewType.Overlay
    } as ExtensionViewDialogState;
    const expectedOverlayFrameSize = {
      width: 640,
      height: 480,
    };
    const instance = wrapper.instance() as RigComponent;
    const frameSize = instance.getFrameSizeFromDialog(overlayTestDialogState);
    expect(frameSize).toEqual(expectedOverlayFrameSize);
  });

  it('returns correct data for custom size', () => {
    const { wrapper } = setupShallow();
    const overlayTestDialogState = {
      width: 100,
      height: 100,
      frameSize: 'Custom',
      extensionViewType: ExtensionViewType.Overlay
    } as ExtensionViewDialogState;
    const expectedOverlayFrameSize = {
      width: 100,
      height: 100,
    };
    const instance = wrapper.instance() as RigComponent;
    const frameSize = instance.getFrameSizeFromDialog(overlayTestDialogState);
    expect(frameSize).toEqual(expectedOverlayFrameSize);
  });

  it('correctly fetches user info if login not in localStorage', () => {
    globalAny.fetch = jest.fn().mockImplementation(mockFetchForUserInfo);
    globalAny.window.location.hash = 'access_token=test&';

    setupShallow();
    expect(globalAny.fetch).toHaveBeenCalled();
  });

  globalAny.fetch = jest.fn().mockImplementation(() => Promise.resolve({
    status: 200,
    json: () => Promise.resolve({}),
  }));

  it('creates project', async () => {
    const { wrapper } = setupShallow();
    const instance = wrapper.instance() as RigComponent;
    const extensionViews = createViewsForTest(1, ExtensionAnchors[ExtensionAnchor.Panel], ViewerTypes.LoggedOut);
    await instance.createProject({ extensionViews } as RigProject);
    expect(globalAny.fetch).toHaveBeenCalled();
  });

  it('selects project', async () => {
    const { wrapper } = setupShallow();
    const instance = wrapper.instance() as RigComponent;
    await instance.selectProject(1);
    expect(globalAny.fetch).toHaveBeenCalled();
  });
});
