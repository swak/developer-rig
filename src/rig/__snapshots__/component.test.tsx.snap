// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RigComponent /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RigComponent
    saveManifest={[Function]}
    session={
        Object {
            "authToken": "test",
            "displayName": "test",
            "id": "test",
            "login": "test",
            "profileImageUrl": "test.png",
          }
    }
    userLogin={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(RigNavComponent)
          createNewProject={[Function]}
          currentProjectIndex={0}
          error={undefined}
          manifest={undefined}
          projects={
                    Array [
                              Object {},
                              Object {},
                            ]
          }
          selectProject={[Function]}
          selectedView="Extension Views"
          viewerHandler={[Function]}
/>,
        <div>
          <ExtensionViewContainer
                    deleteExtensionViewHandler={[Function]}
                    extensionViews={undefined}
                    isLocal={undefined}
                    manifest={undefined}
                    openEditViewHandler={[Function]}
                    openExtensionViewHandler={[Function]}
                    secret={undefined}
          />
          <Console />
</div>,
      ],
      "className": "rig-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "createNewProject": [Function],
          "currentProjectIndex": 0,
          "error": undefined,
          "manifest": undefined,
          "projects": Array [
            Object {},
            Object {},
          ],
          "selectProject": [Function],
          "selectedView": "Extension Views",
          "viewerHandler": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ExtensionViewContainer
              deleteExtensionViewHandler={[Function]}
              extensionViews={undefined}
              isLocal={undefined}
              manifest={undefined}
              openEditViewHandler={[Function]}
              openExtensionViewHandler={[Function]}
              secret={undefined}
/>,
            false,
            false,
            <Console />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "deleteExtensionViewHandler": [Function],
              "extensionViews": undefined,
              "isLocal": undefined,
              "manifest": undefined,
              "openEditViewHandler": [Function],
              "openExtensionViewHandler": [Function],
              "secret": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(RigNavComponent)
            createNewProject={[Function]}
            currentProjectIndex={0}
            error={undefined}
            manifest={undefined}
            projects={
                        Array [
                                    Object {},
                                    Object {},
                                  ]
            }
            selectProject={[Function]}
            selectedView="Extension Views"
            viewerHandler={[Function]}
/>,
          <div>
            <ExtensionViewContainer
                        deleteExtensionViewHandler={[Function]}
                        extensionViews={undefined}
                        isLocal={undefined}
                        manifest={undefined}
                        openEditViewHandler={[Function]}
                        openExtensionViewHandler={[Function]}
                        secret={undefined}
            />
            <Console />
</div>,
        ],
        "className": "rig-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "createNewProject": [Function],
            "currentProjectIndex": 0,
            "error": undefined,
            "manifest": undefined,
            "projects": Array [
              Object {},
              Object {},
            ],
            "selectProject": [Function],
            "selectedView": "Extension Views",
            "viewerHandler": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ExtensionViewContainer
                deleteExtensionViewHandler={[Function]}
                extensionViews={undefined}
                isLocal={undefined}
                manifest={undefined}
                openEditViewHandler={[Function]}
                openExtensionViewHandler={[Function]}
                secret={undefined}
/>,
              false,
              false,
              <Console />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "deleteExtensionViewHandler": [Function],
                "extensionViews": undefined,
                "isLocal": undefined,
                "manifest": undefined,
                "openEditViewHandler": [Function],
                "openExtensionViewHandler": [Function],
                "secret": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
            false,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<RigComponent /> renders extension view correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RigComponent
    saveManifest={[Function]}
    session={
        Object {
            "authToken": "test",
            "displayName": "test",
            "id": "test",
            "login": "test",
            "profileImageUrl": "test.png",
          }
    }
    userLogin={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(RigNavComponent)
          createNewProject={[Function]}
          currentProjectIndex={0}
          error={undefined}
          manifest={undefined}
          projects={
                    Array [
                              Object {
                                "extensionViews": Array [
                                  Object {
                                    "channelId": "twitch",
                                    "features": Object {
                                      "isChatEnabled": true,
                                    },
                                    "id": "1",
                                    "isPopout": false,
                                    "linked": false,
                                    "linkedUserId": "",
                                    "mode": "viewer",
                                    "opaqueId": "ARIGopaqueId",
                                    "orientation": "Portrait",
                                    "role": "Logged-Out Viewer",
                                    "type": "Panel",
                                    "x": 0,
                                    "y": 0,
                                  },
                                ],
                              },
                            ]
          }
          selectProject={[Function]}
          selectedView="Extension Views"
          viewerHandler={[Function]}
/>,
        <div>
          <ExtensionViewContainer
                    deleteExtensionViewHandler={[Function]}
                    extensionViews={
                              Array [
                                        Object {
                                          "channelId": "twitch",
                                          "features": Object {
                                            "isChatEnabled": true,
                                          },
                                          "id": "1",
                                          "isPopout": false,
                                          "linked": false,
                                          "linkedUserId": "",
                                          "mode": "viewer",
                                          "opaqueId": "ARIGopaqueId",
                                          "orientation": "Portrait",
                                          "role": "Logged-Out Viewer",
                                          "type": "Panel",
                                          "x": 0,
                                          "y": 0,
                                        },
                                      ]
                    }
                    isLocal={undefined}
                    manifest={undefined}
                    openEditViewHandler={[Function]}
                    openExtensionViewHandler={[Function]}
                    secret={undefined}
          />
          <Console />
</div>,
      ],
      "className": "rig-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "createNewProject": [Function],
          "currentProjectIndex": 0,
          "error": undefined,
          "manifest": undefined,
          "projects": Array [
            Object {
              "extensionViews": Array [
                Object {
                  "channelId": "twitch",
                  "features": Object {
                    "isChatEnabled": true,
                  },
                  "id": "1",
                  "isPopout": false,
                  "linked": false,
                  "linkedUserId": "",
                  "mode": "viewer",
                  "opaqueId": "ARIGopaqueId",
                  "orientation": "Portrait",
                  "role": "Logged-Out Viewer",
                  "type": "Panel",
                  "x": 0,
                  "y": 0,
                },
              ],
            },
          ],
          "selectProject": [Function],
          "selectedView": "Extension Views",
          "viewerHandler": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ExtensionViewContainer
              deleteExtensionViewHandler={[Function]}
              extensionViews={
                            Array [
                                          Object {
                                            "channelId": "twitch",
                                            "features": Object {
                                              "isChatEnabled": true,
                                            },
                                            "id": "1",
                                            "isPopout": false,
                                            "linked": false,
                                            "linkedUserId": "",
                                            "mode": "viewer",
                                            "opaqueId": "ARIGopaqueId",
                                            "orientation": "Portrait",
                                            "role": "Logged-Out Viewer",
                                            "type": "Panel",
                                            "x": 0,
                                            "y": 0,
                                          },
                                        ]
              }
              isLocal={undefined}
              manifest={undefined}
              openEditViewHandler={[Function]}
              openExtensionViewHandler={[Function]}
              secret={undefined}
/>,
            false,
            false,
            <Console />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "deleteExtensionViewHandler": [Function],
              "extensionViews": Array [
                Object {
                  "channelId": "twitch",
                  "features": Object {
                    "isChatEnabled": true,
                  },
                  "id": "1",
                  "isPopout": false,
                  "linked": false,
                  "linkedUserId": "",
                  "mode": "viewer",
                  "opaqueId": "ARIGopaqueId",
                  "orientation": "Portrait",
                  "role": "Logged-Out Viewer",
                  "type": "Panel",
                  "x": 0,
                  "y": 0,
                },
              ],
              "isLocal": undefined,
              "manifest": undefined,
              "openEditViewHandler": [Function],
              "openExtensionViewHandler": [Function],
              "secret": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(RigNavComponent)
            createNewProject={[Function]}
            currentProjectIndex={0}
            error={undefined}
            manifest={undefined}
            projects={
                        Array [
                                    Object {
                                      "extensionViews": Array [
                                        Object {
                                          "channelId": "twitch",
                                          "features": Object {
                                            "isChatEnabled": true,
                                          },
                                          "id": "1",
                                          "isPopout": false,
                                          "linked": false,
                                          "linkedUserId": "",
                                          "mode": "viewer",
                                          "opaqueId": "ARIGopaqueId",
                                          "orientation": "Portrait",
                                          "role": "Logged-Out Viewer",
                                          "type": "Panel",
                                          "x": 0,
                                          "y": 0,
                                        },
                                      ],
                                    },
                                  ]
            }
            selectProject={[Function]}
            selectedView="Extension Views"
            viewerHandler={[Function]}
/>,
          <div>
            <ExtensionViewContainer
                        deleteExtensionViewHandler={[Function]}
                        extensionViews={
                                    Array [
                                                Object {
                                                  "channelId": "twitch",
                                                  "features": Object {
                                                    "isChatEnabled": true,
                                                  },
                                                  "id": "1",
                                                  "isPopout": false,
                                                  "linked": false,
                                                  "linkedUserId": "",
                                                  "mode": "viewer",
                                                  "opaqueId": "ARIGopaqueId",
                                                  "orientation": "Portrait",
                                                  "role": "Logged-Out Viewer",
                                                  "type": "Panel",
                                                  "x": 0,
                                                  "y": 0,
                                                },
                                              ]
                        }
                        isLocal={undefined}
                        manifest={undefined}
                        openEditViewHandler={[Function]}
                        openExtensionViewHandler={[Function]}
                        secret={undefined}
            />
            <Console />
</div>,
        ],
        "className": "rig-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "createNewProject": [Function],
            "currentProjectIndex": 0,
            "error": undefined,
            "manifest": undefined,
            "projects": Array [
              Object {
                "extensionViews": Array [
                  Object {
                    "channelId": "twitch",
                    "features": Object {
                      "isChatEnabled": true,
                    },
                    "id": "1",
                    "isPopout": false,
                    "linked": false,
                    "linkedUserId": "",
                    "mode": "viewer",
                    "opaqueId": "ARIGopaqueId",
                    "orientation": "Portrait",
                    "role": "Logged-Out Viewer",
                    "type": "Panel",
                    "x": 0,
                    "y": 0,
                  },
                ],
              },
            ],
            "selectProject": [Function],
            "selectedView": "Extension Views",
            "viewerHandler": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ExtensionViewContainer
                deleteExtensionViewHandler={[Function]}
                extensionViews={
                                Array [
                                                Object {
                                                  "channelId": "twitch",
                                                  "features": Object {
                                                    "isChatEnabled": true,
                                                  },
                                                  "id": "1",
                                                  "isPopout": false,
                                                  "linked": false,
                                                  "linkedUserId": "",
                                                  "mode": "viewer",
                                                  "opaqueId": "ARIGopaqueId",
                                                  "orientation": "Portrait",
                                                  "role": "Logged-Out Viewer",
                                                  "type": "Panel",
                                                  "x": 0,
                                                  "y": 0,
                                                },
                                              ]
                }
                isLocal={undefined}
                manifest={undefined}
                openEditViewHandler={[Function]}
                openExtensionViewHandler={[Function]}
                secret={undefined}
/>,
              false,
              false,
              <Console />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "deleteExtensionViewHandler": [Function],
                "extensionViews": Array [
                  Object {
                    "channelId": "twitch",
                    "features": Object {
                      "isChatEnabled": true,
                    },
                    "id": "1",
                    "isPopout": false,
                    "linked": false,
                    "linkedUserId": "",
                    "mode": "viewer",
                    "opaqueId": "ARIGopaqueId",
                    "orientation": "Portrait",
                    "role": "Logged-Out Viewer",
                    "type": "Panel",
                    "x": 0,
                    "y": 0,
                  },
                ],
                "isLocal": undefined,
                "manifest": undefined,
                "openEditViewHandler": [Function],
                "openExtensionViewHandler": [Function],
                "secret": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            false,
            false,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
